### ISSUE TITLE
Server waits to send last segment until all ACKs for previous segments are received.

### ISSUE THREAD

**Issue by micgolde** [cisco-system-traffic-generator/trex-core#436]
My server is waiting for all ACKs before sending it's last segment (in my case a HTTP 200 OK).

The client also appears to be hitting the delayed ack timer (200ms) before finally sending the last ACK of the second to last segment.  There isn't any reason why the server shouldn't be sending the 200 OK (last segment) in this situation.

PCAP (time-stamps on this are somewhat unreliable due to the box I have in the middle, but they're in the right order)
[TR1024B_SingleConnection_DAfix.zip](https://github.com/cisco-system-traffic-generator/trex-core/files/4220925/TR1024B_SingleConnection_DAfix.zip)

def profile(**kwargs):
    http_req = 'GET /1028 HTTP/1.1\r\nHost: 10.10.0.1\r\nConnection: close\r\nUser-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1) Opera 8.0\r\nAccept: */*\r\nAccept-Language: en-us\r\nAccept-Encoding: gzip, deflate, compress\r\n\r\n'
    http_response = 'HTTP/1.1 200 OK\r\nServer: Jetty/4.2.9rc2 (SunOS/5.8 sparc java/1.4.1_04)\r\nContent-Type: text/html\r\nConnection: close\r\nContent-Length: 256000\r\n\r\n'
    http_response2 = 255976 * '*'
    http_total_response = http_response + '<html><pre>' + http_response2 + '</pre></html>'

    # client commands
    prog_c = trex.astf.trex_astf_profile.ASTFProgram()
    prog_c.connect();
    prog_c.send(http_req)
    prog_c.recv(len(http_total_response))
    prog_c.wait_for_peer_close()

    prog_s = trex.astf.trex_astf_profile.ASTFProgram()
    prog_s.recv(len(http_req))
    prog_s.send(http_total_response)

    info = trex.astf.trex_astf_profile.ASTFGlobalInfo()
    info.tcp.mss = 1460
    info.tcp.rxbufsize = 32768
    info.tcp.txbufsize = 32768
    info.tcp.initwnd = 1
    info.tcp.delay_ack_msec = 200
    info.tcp.no_delay_counter = 2
    info.tcp.do_rfc1323 = 0

    return {'prog_c': prog_c, 'prog_s': prog_s, 'info': info}

**Comment by hhaim** [comment]
@micgolde try adding set_blocking mode to false

**Comment by micgolde** [comment]
That appears to completely break it... I'm not sure what it's doing, but it doesn't appear to be closing the connection properly at all anymore:

    prog_s = trex.astf.trex_astf_profile.ASTFProgram()
    prog_s.set_send_blocking (False)
    prog_s.recv(len(http_req))
    prog_s.send(http_total_response)

PCAP:
[blocking_mode_false.zip](https://github.com/cisco-system-traffic-generator/trex-core/files/4222227/blocking_mode_false.zip)

**Comment by hhaim** [comment]
Try adding this in the client side

**Comment by micgolde** [comment]
That still doesn't help.  It's still waiting on all the acks from all the segments before sending the last segment:

    prog_c = trex.astf.trex_astf_profile.ASTFProgram()
    prog_c.connect();
    prog_c.set_send_blocking (False)
    prog_c.send(http_req)
    prog_c.recv(len(http_total_response))

    prog_s = trex.astf.trex_astf_profile.ASTFProgram()
    prog_s.recv(len(http_req))
    prog_s.send(http_total_response)

PCAP:
[client_side_blocking.zip](https://github.com/cisco-system-traffic-generator/trex-core/files/4233443/client_side_blocking.zip)

**Comment by hhaim** [comment]
I don't see this  
>>My server is waiting for all ACKs before sending it's last segment (in my case a HTTP 200 OK)

[http_436.py.tar.gz](https://github.com/cisco-system-traffic-generator/trex-core/files/4437273/http_436.py.tar.gz)
[a3.pcap.tar.gz](https://github.com/cisco-system-traffic-generator/trex-core/files/4437281/a3.pcap.tar.gz)

The command I'm using to run the simulation 
```./astf-sim -f astf/http_436.py -o a3.pcap --full```

```prog_c.wait_for_peer_close()``` is only valid in the server side program not the client side.
Wanted to be sure if we are talking about a delay with response or a feature request 
delay with response is a bug and we should fix it. 

You can do delay+reset to the server side
```
        prog_c = ASTFProgram()
        prog_c.connect();
        prog_c.send(http_req)
        prog_c.recv(len(http_total_response))
        prog_c.delay(1000)


        prog_s = ASTFProgram()
        prog_s.recv(len(http_req))
        prog_s.send(http_total_response)
        prog_s.reset()

```

**Comment by hhaim** [comment]
@micgolde I see that you have another issue #435 (to add a feature to close by the server). So could you say again what is this issue about

**Comment by micgolde** [comment]
This isn't directly related to #435.  #435 I think will just end up being a feature request where I can tell the server to close the connection (similar to the reset command, but close).

This bug is actually shown in the pcap you shared too:
![image](https://user-images.githubusercontent.com/53269288/78564182-0baa8a80-77ea-11ea-8dfe-d734c0594f9b.png)

trex waits to send the last packet (packet 1236) until it receives all acks for previous packets.

This gets much worse when you add delayed ack, since the client is waiting to ack and the server isn't sending anymore packets.  In my pcap you can see that there is almost a 0.18 second delay between packets 266 and 267, this is the delayed ack timer timing out waiting for the server to send the last packet:

![image](https://user-images.githubusercontent.com/53269288/78564083-f0d81600-77e9-11ea-9b88-5a90aa4bfa20.png)

**Comment by hhaim** [comment]
@micgolde #435 is related to this issue and both has the same root cause. 
Will provide a patch that solves this. Thanks for identifying this.
